beacon-chain/sync/data_column_sidecars.go

```diff
@@ -300,80 +445,202 @@ func tryRequestingColumnsFromPeers(
-			verifiedColumnsByRoot[root] = append(verifiedColumnsByRoot[root], verifiedRoDataColumns...)
+			result[root] = append(result[root], verifiedRoDataColumns...)
+		}
+
+		// Unlabel a root as to retrieve if enough sidecars are retrieved to enable a reconstruction,
+		// or if all requested sidecars are now available for this root.
+		for root, indicesToRetrieve := range indicesToRetrieveByRoot {
+			storedIndices := storedIndicesByRoot[root]
+			storedCount := uint64(len(storedIndices))
+			resultCount := uint64(len(result[root]))
+
+			if storedCount+resultCount >= minimumColumnCountToReconstruct {
+				delete(indicesToRetrieveByRoot, root)
+				continue
+			}
+
+			allRequestedIndicesAvailable := true
+			for index := range requestedIndices {
+				if indicesToRetrieve[index] {
+					// Still need this index.
+					allRequestedIndicesAvailable = false
+					break
+				}
+			}
+
+			if allRequestedIndicesAvailable {
+				delete(indicesToRetrieveByRoot, root)
+			}
 		}
 
 		// Compute indices by root by peers with the updated missing indices and connected peers.
-		indicesByRootByPeer, err = computeIndicesByRootByPeer(p.P2P, slotByRoot, missingIndicesByRoot, connectedPeers)
+		indicesByRootByPeer, err = computeIndicesByRootByPeer(p.P2P, slotByRoot, indicesToRetrieveByRoot, connectedPeers)
 		if err != nil {
 			return nil, errors.Wrap(err, "explore peers")
 		}
 	}
 
-	return verifiedColumnsByRoot, nil
+	log.WithFields(logrus.Fields{
+		"duration":                   time.Since(start),
+		"initialToRetrieveRootCount": initialToRetrieveRootCount,
+		"finalToRetrieveRootCount":   len(indicesToRetrieveByRoot),
+	}).Debug("Requested all data column sidecars from peers")
+
+	return result, nil
 }
 
-// tryReconstructFromStorageAndPeers attempts to reconstruct missing data column sidecars
-// using the data available in the storage and the data fetched from peers.
-// If, for at least one root, the reconstruction is not possible, an error is returned.
-func tryReconstructFromStorageAndPeers(
+// mergeStorageAndInputs retrieves missing data column sidecars by combining
+// what is available in storage with the sidecars provided in `alreadyAvailableByRoot`,
+// reconstructing them when necessary.
+//
+// The function works in two modes depending on sidecar availability:
+//   - If all requested sidecars are already available (no reconstruction needed),
+//     it simply returns them directly from storage and inputs.
+//   - If storage + inputs together provide enough sidecars to reconstruct all requested ones,
+//     it reconstructs and returns the requested sidecars.
+//
+// If a root cannot yield all requested sidecars, that root is omitted from the result.
+//
+// Note: It is assumed that no sidecar in `alreadyAvailableByRoot` is already present in storage.
+//
+// WARNING: This function mutates `roots`, removing any block roots
+// for which all requested sidecars were successfully retrieved.
+func mergeStorageAndInputs(
```

Comment:
what about mergeAvailableSidecars instead?

Reply:
Fixed in 3b51563bca57dcc194cbcdf45dbb2ac73028d286.